leshan-client-cf/src/main/java/org/eclipse/leshan/client/californium/DefaultLeshanCertificateVerifier.java

```diff
@@ -0,0 +1,212 @@
+            X509Certificate[] trustedCertificates, CertificateUsage certificateUsage) {
+        this.expectedServerCertificate = expectedServerCertificate;
+        this.trustedCertificates = trustedCertificates;
+        this.certificateUsage = certificateUsage;
+    }
+
+    @Override
+    public CertPath verifyCertificate(Boolean clientUsage, boolean truncateCertificatePath,
+            CertificateMessage message, DTLSSession session) throws HandshakeException {
+        CertPath messageChain = message.getCertificateChain();
+
+        if (messageChain.getCertificates().size() == 0) {
+            AlertMessage alert = new AlertMessage(AlertLevel.FATAL, AlertDescription.BAD_CERTIFICATE,
+                    session.getPeer());
+            throw new HandshakeException(
+                    "Certificate chain could not be validated : server cert chain is empty", alert);
+        }
+
+        Certificate receivedServerCertificate = messageChain.getCertificates().get(0);
+        if (!(receivedServerCertificate instanceof X509Certificate)) {
+            AlertMessage alert = new AlertMessage(AlertLevel.FATAL, AlertDescription.UNSUPPORTED_CERTIFICATE,
+                    session.getPeer());
+            throw new HandshakeException(
+                    "Certificate chain could not be validated - unknown certificate type", alert);
+        }
+        X509Certificate serverCertificate = (X509Certificate) receivedServerCertificate;
+
+        // If clientUsage is defined then check key usage
+        if (clientUsage) {
+            if (!CertPathUtil.canBeUsedForAuthentication(serverCertificate, true)) {
+                AlertMessage alert = new AlertMessage(AlertLevel.FATAL, AlertDescription.BAD_CERTIFICATE,
+                        session.getPeer());
+                throw new HandshakeException("Certificate chain could not be validated - Key Usage doesn't match!",
+                        alert);
+            }
+        }
+
+        CertPath certPath = expandCertPath(messageChain);
+        if (trustedCertificates != null) {
+            try {
+                // - must do PKIX validation with trustStore
+                certPath = CertPathUtil
+                        .validateCertificatePath(false, certPath, trustedCertificates);
+            } catch (GeneralSecurityException e) {
+                AlertMessage alert = new AlertMessage(AlertLevel.FATAL, AlertDescription.BAD_CERTIFICATE,
+                        session.getPeer());
+                throw new HandshakeException("Certificate chain could not be validated", alert, e);
+            }
+        }
+
+        if (certificateUsage == CertificateUsage.CA_CONSTRAINT) {
+            boolean found = false;
+
+            // - must check that given certificate is part of certPath
+            List<? extends Certificate> certificates = certPath.getCertificates();
+            for (Certificate certificate : certificates) {
+                if (certificate.equals(expectedServerCertificate)) {
```

Comment:
> In what later check do you refer here?

Checks like here. Therefore I mentioned the `expectedServerCertificate`in my comment and that should have made it easy to see, where it is used.

Reply:
Let's assume following scenario:
- LWM2M installation would be using public PKI
- From public PKI there would be intermediate CA in control of <company X> while still hosted by public PKI vendor
- LWM2M Bootstrap server would have issued certificate under that intermediate CA
- LWM2M Client trust store would be common CA certificates so that it could connect to any valid bootstrap server
- LWM2M Client connects to LWM2M Bootstrap server
- LWM2M Client will be given CA constraint usage and then installations intermediate CA
- Now when LWM2M Client connects to LWM2M Server there can still be many CA's in the store AND also the given intermediate CA but in order to constraint that to particular <company X> then CA constraint limits it to that part of the tree

Or there could be private CA setup and and it is still wanted to have more restrictions by operators wish. eg. limit development devices to only testing environment or so.

Point is to implement it as specified and let the operator decide what to do so that there are no surprises.